dash/dash.py

```diff
@@ -2076,3 +2166,136 @@ def verify_url_part(served_part, url_part, part_name):
+                page_filename = page_filename.replace(".py", "").replace("/", ".")
+                module_name = ".".join([pages, page_filename])
+                page_module = importlib.import_module(module_name)
+
+                if not _pages.PAGE_REGISTRY[module_name]["supplied_layout"]:
+                    _validate.validate_pages_layout(module_name, page_module)
+                    _pages.PAGE_REGISTRY[module_name]["layout"] = getattr(
+                        page_module, "layout"
+                    )
+
+    @staticmethod
+    def _path_to_page(path_id):
+        path_variables = None
+        for page in _pages.PAGE_REGISTRY.values():
+            if page["path_template"]:
+                template_id = page["path_template"].strip("/")
+                path_variables = _parse_path_variables(path_id, template_id)
+                if path_variables:
+                    return page, path_variables
+            if path_id == page["path"].strip("/"):
+                return page, path_variables
+        return {}, None
+
+    def enable_pages(self):
+        if not self.use_pages:
+            return
+        if self.pages_folder and not os.path.exists(self.config.pages_folder):
+            warnings.warn(
+                f"A folder called {self.pages_folder} does not exist.", stacklevel=2
+            )
+        if os.path.exists(self.config.pages_folder):
+            self._import_layouts_from_pages()
+
+        @self.server.before_first_request
+        def router():
+            @self.callback(
+                Output(_ID_CONTENT, "children"),
+                Output(_ID_STORE, "data"),
+                Input(_ID_LOCATION, "pathname"),
+                Input(_ID_LOCATION, "search"),
+                prevent_initial_call=True,
+            )
+            def update(pathname, search):
+                """
+                Updates dash.page_container layout on page navigation.
+                Updates the stored page title which will trigger the clientside callback to update the app title
+                """
+
+                query_parameters = _parse_query_string(search)
+                page, path_variables = self._path_to_page(
+                    self.strip_relative_path(pathname)
+                )
+
+                # get layout
+                if page == {}:
+                    module_404 = ".".join([self.pages_folder, "not_found_404"])
+                    not_found_404 = _pages.PAGE_REGISTRY.get(module_404)
+                    if not_found_404:
+                        layout = not_found_404["layout"]
+                        title = not_found_404["title"]
+                    else:
+                        layout = html.H1("404 - Page not found")
+                        title = self.title
+                else:
+                    layout = page["layout"]
+                    title = page["title"]
+
+                if isinstance(layout, patch_collections_abc("Callable")):
```

Comment:
Despite the fact that we have this pattern elsewhere in this repo, `callable(layout)` like you [have below](https://github.com/plotly/dash/pull/1947/files#diff-187d1883eb8273e47c7fae609bb5667d5a9488f9406dfc6a9ca5f19d8d7cc205R2268) is a [better pattern](https://stackoverflow.com/questions/624926/how-do-i-detect-whether-a-python-variable-is-a-function)

(and [6 lines below](https://github.com/plotly/dash/pull/1947/files#diff-187d1883eb8273e47c7fae609bb5667d5a9488f9406dfc6a9ca5f19d8d7cc205R2258) as well)